crl/validator.go

```diff
@@ -20,346 +20,420 @@ package crl
+			}
+			// add any previously unknown CRL distribution points.
+			// (Leaf cert is likely to contain CRLDistributionPoints not found in the truststore.)
+			if len(cert.CRLDistributionPoints) > 0 {
+				for _, endpoint := range cert.CRLDistributionPoints {
+					if v.crls[endpoint] == nil {
+						v.crls[endpoint] = newRevocationList(issuer)
+					}
+				}
+			}
+		}
 	}
 
-	v.listsLock.Lock()
-	defer v.listsLock.Unlock()
+	config.VerifyPeerCertificate = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// TODO: the truststore should already contain ALL CAs that might be found in the verifiedChains
+		// TODO: should we check rawCerts, or verified chains?
+		// rawCerts contains raw cert data from the tls handshake.
+		// It is used together with the tls.Config to generate verifiedChains, but may contain additional certs that are not in a verified chain.
 
-	issuerName := crl.Issuer.String()
+		certificates, err := parseCertificatesSlice(rawCerts)
+		if err != nil {
+			return err
+		}
 
-	for _, cert := range crl.RevokedCertificates {
-		v.setRevoked(issuerName, cert.SerialNumber)
+		return v.validateChain(certificates)
 	}
-
-	v.lists[endpoint] = crl
-
 	return nil
 }
 
-// crlHasExpired is a replacement for pkix.CertificateList.HasExpired
-func crlHasExpired(crl *x509.RevocationList, deadline time.Time) bool {
-	return !deadline.Before(crl.NextUpdate)
+// validateChain validates that none of the certificates in the chain is revoked.
+// Certificated in chain are assumed ordered from leaf to root certificate.
+func (v *validator) validateChain(chain []*x509.Certificate) error {
+	var cert *x509.Certificate
+	var err error
+	for i := range chain {
+		cert = chain[len(chain)-1-i]
+		// check in reverse order to prevent CRL expiration errors due to revoked CAs no longer issuing CRLs
+		if err = v.validateCert(cert); err != nil {
+			return fmt.Errorf("%w: subject=%s, S/N=%s, issuer=%s", err, cert.Subject.String(), cert.SerialNumber.String(), cert.Issuer.String())
+		}
+	}
+	return nil
 }
 
-// IsSynced returns whether all the CRLs are downloaded and are not outdated (based on the offset)
-func (v *validator) IsSynced(maxOffsetDays int) bool {
-	endpoints := v.parseCRLEndpoints()
-
-	v.listsLock.RLock()
-	defer v.listsLock.RUnlock()
-
-	// Check if all CRLs have been downloaded, but if ignore CRL endpoints that are only used by expired certificates.
-	for endpoint, dependingCAs := range endpoints {
-		if !anyCertificateActive(dependingCAs) {
-			continue
+func (v *validator) validateCert(cert *x509.Certificate) error {
+	if nowFunc().After(cert.NotAfter) {
+		return ErrCertExpired
+	}
+	for _, endpoint := range cert.CRLDistributionPoints {
+		crl, ok := v.getCRL(endpoint)
+		if ok && crl.revoked[cert.SerialNumber.String()] != nil {
+			// revocation takes precedence over an internal error.
+			return ErrCertRevoked
+		}
+		if !ok || crl.lastUpdated.IsZero() {
```

Comment:
this is in case the CRL isn't downloaded yet? It can just fail i.m.o., since the CRLs are updated as the first thing when starting the node?

`Start()` could even block until the inital CRL download has finished (with a sane time-out, of course, to avoid blocking the node bootstrap).

If we don't block, initial calls might fail (or we block-wait here until the initial download has finished).

Reply:
Yes, the CRL is not downloaded yet.

The issue is that we do not know the all the CRLDistributionPoints at startup. We know those in the truststore, but the leaf certificates contain some missing CRL endpoints (the TSPs CRL endpoints). We have 1 of them from our own cert, but the pkiO chain has 5 TSPs, so some of them we just need to encounter at runtime.